### ISSUE TITLE
[RFC]: Truncate mode for `thinking_token_budget`

### ISSUE THREAD

**Issue by walterbm-cohere** [vllm-project/vllm#54864]
### Motivation.

`thinking_token_budget` currently enforces the budget by forcing the logits of
`reasoning_end_str` (e.g. `</think>`) once the reasoning block reaches the budget, then
letting the model continue with the answer. That make sense for general serving but is not ideal for RL
rollouts:

- The forced tokens are not sampled from the policy, so they are off-policy tokens in the
  trajectory and their logprobs are polluted by the 1e9 logit override.
- There is no signal that the budget was hit; the response finishes with the usual
  `finish_reason="stop"`.
- RL trainers typically want an over-budget sample to end right there, be marked as
  truncated (`finish_reason="length"`), and free its KV cache instead of generating an answer

### Proposed Change.

Add an engine-wide flag to `ReasoningConfig`:

```python
thinking_budget_action: Literal["force_end", "truncate"] = "force_end" # keeps current behavior as default
```

```bash
vllm serve ... --reasoning-config '{ ... "thinking_budget_action": "truncate"}'
```

With `thinking_budget_action=truncate`:

- The budget is enforced in the scheduler, not the sampler. A small tracker on
  `Request` (reasoning start position, reasoning token count) is updated per output token
  in `_update_request_with_output`, using the start and natural-end token ids already
  derived by `ReasoningConfig.initialize_token_ids`. Prompt-resident reasoning tokens
  are counted at request creation, as the current implementation does.
- `check_stop` add one conditional branch next to the `max_tokens` check: when the count reaches
  the budget the request is finished with a new `RequestStatus.FINISHED_THINKING_BUDGET`,
  mapped to `FinishReason.LENGTH`, with `stop_reason="thinking_token_budget"`. No new
  `FinishReason` value, so API and Rust DTOs are unchanged.
- The sampler-side engines are disabled at construction in this mode
  (`maybe_create_thinking_budget_state_holder` returns `None`; V2 `ThinkingBudgetState`
  is constructed disabled). No token is ever forced.
- Speculative decoding and async scheduling need no special handling: the existing
  per-token stop loop already trims tokens after a stop, and the one-step lag is the same as
  for `max_tokens`.

List of affected files (probably): `vllm/config/reasoning.py`, `vllm/v1/request.py`,
`vllm/v1/core/sched/utils.py`, one condition each in `gpu_input_batch.py` and
`gpu/sample/thinking_budget.py`, docs in `reasoning_outputs.md`, unit tests in
`tests/v1/core` and an e2e case in `test_thinking_token_budget.py`.

### Any Other Things.

🤚 happy to contribute this feature if accepted! I already have a working implementation 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by chaunceyjiang** [comment]
@walterbm-cohere  How is `thinking_budget_action="truncate"` materially different from `max_tokens` for the intended RL rollout use case? If reasoning always starts at a known position, it  seems the caller could set `max_tokens` to the reasoning budget.

**Comment by walterbm-cohere** [comment]
> [@walterbm-cohere](https://github.com/walterbm-cohere) How is `thinking_budget_action="truncate"` materially different from `max_tokens` for the intended RL rollout use case? If reasoning always starts at a known position, it seems the caller could set `max_tokens` to the reasoning budget.

@chaunceyjiang the difference is that in the RL use-case we don't want to cap the answer length. just the reasoning. if we set `max_tokens=thinking_budget` we model would be stopped short during the answer generation when reasoning is <  thinking_budget

**Comment by aoshen02** [comment]
Design looks solid overall. One question: the existing `force_end` code gives natural `</think>` priority over the budget countdown when both hit on the same step — does the new `check_stop` logic preserve that tie-break, or could a normal completion get mislabeled as truncated? Worth a test for that boundary case either way.

**Comment by de1tydev** [comment]
One serving-side note to keep the two modes apart: with the default `force_end`, the case that bites OpenAI-compatible clients is not the budget being hit — it is the budget never being set while `max_tokens` is small, so the request ends with `finish_reason="length"` and empty `content` before any answer. We run a server-side clamp (`thinking_token_budget = min(client max_tokens, effective max_tokens) − reserve` when the client sets no budget) and it turned a 145-case strict tool-calling gate from 1–11 failures/run into 0/145 with no benchmark cost; details and numbers in #55382. Just flagging that `truncate` should stay opt-in — for serving, `force_end` plus a reserve relative to `max_tokens` is the behaviour clients expect.

**Comment by aoshen02** [comment]
Good catch, will fix it!

**Comment by BugenZhao** [comment]
Since we already report reasoning token counts in the usage stats, could the client simply cancel the request once the budget is reached? Is there a limitation with that approach that requires engine-side enforcement?
